proposals/service_api_calls.md

```diff
@@ -0,0 +1,186 @@
+
+Finally, in some scenarios 3rd party library code or binaries need to be executed to make a policy decision. Currently, the only way to support these "extensions" is to directly update the Kyverno core. With this feature, it will be possible to write Kyverno extensions as separate services hence decoupling the Kyverno code from 3rd party libraries or binaries.
+
+# Proposal
+
+This feature enables:
+- HTTP GET and POST API calls to any service
+- Encryption and server authentication using a certificate bundle
+- Client authentication using Service Account token volume projection 
+- Passing context data to the external service
+- Applying JMESPath transformation to the results
+- Storing service API call results in the rule context for further processing
+
+# Implementation
+
+Here are the changes involved:
+1. Add a new `service` to the existing `APICall` declaration, to allow specifying information like the URL and CA bundle. The existing `
+2. Update the logic to handle APICall to check whether a `urlPath` is configured to invoke the API server. Otherwise, process the `service` declaration to execute an HTTP GET or POST to the JSON web service.
+3. Apply the JMESPath (if specified) to the results and store in context, as before. 
+
+
+## CRD Changes
+
+Changes for the APICall object:
+
+```sh
+ ❯ kubectl explain cpol.spec.rules.context.apiCall
+
+KIND:     ClusterPolicy
+VERSION:  kyverno.io/v1
+
+RESOURCE: apiCall <Object>
+
+DESCRIPTION:
+     APICall is an HTTP request to the Kubernetes API server, or other JSON web
+     service. The data returned is stored in the context with the name for the
+     context entry.
+
+FIELDS:
+   jmesPath     <string>
+     JMESPath is an optional JSON Match Expression that can be used to transform
+     the JSON response returned from the server. For example a JMESPath of
+     "items | length(@)" applied to the API server response for the URLPath
+     "/apis/apps/v1/deployments" will return the total count of deployments
+     across all namespaces.
+
+   service      <Object>
+     Service is an API call to a JSON web service
+
+   urlPath      <string>
+     URLPath is the URL path to be used in the HTTP GET request to the
+     Kubernetes API server (e.g. "/api/v1/namespaces" or
+     "/apis/apps/v1/deployments"). The format required is the same format used
+     by the `kubectl get --raw` command.
+
+```
+
+New ServiceCall object:
+
+```sh
+❯ kubectl explain cpol.spec.rules.context.apiCall.service
+
+KIND:     ClusterPolicy
+VERSION:  kyverno.io/v1
+
+RESOURCE: service <Object>
+
+DESCRIPTION:
+     Service is an API call to a JSON web service
+
+FIELDS:
+   caBundle     <string>
+     CABundle is a PEM encoded CA bundle which will be used to validate the
+     server certificate.
+
+   data <[]Object>
```

Comment:
Can we also add a field to specificy URL query parameters in addition to post/body? It will make the user experience easy and Kyverno can handle properly constructing/encoding the url params.

Reply:
I've noted in the future enhancements section.